website/src/theme/DocItem/Layout/index.js

```diff
@@ -42,74 +42,77 @@ function useDocTOC() {
   const { version: dbtVersion } = useContext(VersionContext)
   const [currentToc, setCurrentToc] = useState(toc)
   const [tocReady, setTocReady] = useState(true)
-  useEffect(() => {
-    async function fetchElements() {
-      // get html elements
-      const headings = await getElements(".markdown h1, .markdown h2, .markdown h3, .markdown h4, .markdown h5, .markdown h6")
-      
-      // if headings exist on page
-      // compare against toc
-      if(toc && headings && headings.length) {
-        // make new TOC object 
-        let updated = Array.from(headings).reduce((acc, item) => {
-          // If heading id and toc item id match found
-          // include in updated toc
-          let found = toc.find(heading =>
-            heading.id.includes(item.id)
-          )
-          // If toc item is not in headings
-          // do not include in toc
-          // This means heading is versioned
-
-          let makeToc = (heading) => {
-            let level;
-            if (heading.nodeName === "H2") {
-              level = 2
-            } else if (heading.nodeName === "H3") {
-              level = 3
-            } else {
-              level = null
-            }
 
-            return {
-              value: heading.innerHTML,
-              id: heading.id,
-              level: level && level
-            }
-          }
+  async function fetchElements() {
+    // get html elements
+    const headings = await getElements(".markdown h1, .markdown h2, .markdown h3, .markdown h4, .markdown h5, .markdown h6")
+    
+    // if headings exist on page
+    // compare against toc
+    if(toc && headings && headings.length) {
+      // make new TOC object 
+      let updated = Array.from(headings).reduce((acc, item) => {
+        // If heading id and toc item id match found
+        // include in updated toc
+        let found = toc.find(heading =>
+          heading.id.includes(item.id)
+        )
+        // If toc item is not in headings
+        // do not include in toc
+        // This means heading is versioned
 
-          if (found) {
-            acc.push(makeToc(item))
-          } else if (!found) {
-            acc.push(makeToc(item))
+        let makeToc = (heading) => {
+          let level;
+          if (heading.nodeName === "H2") {
+            level = 2
+          } else if (heading.nodeName === "H3") {
+            level = 3
           } else {
-            null
+            level = null
           }
 
-          return acc
-        }, [])
+          return {
+            value: heading.innerHTML,
+            id: heading.id,
+            level: level && level
+          }
+        }
 
-        // If updated toc different than current
-        // If so, show loader and update toc 
-        if(currentToc.length !== updated.length) {
```

Comment:
This logic was flawed! 
Here, we were comparing the `length` of the heading data of the page to add to the TOC (`updated`, which equals an array from all the headings on the page) to the `length` of the TOC data currently being used to render the TOC component `currentToc`.

In some cases, even though `currentToc` and `updated` were different content-wise, they could have the same `length`. `updated` has ALL the headings including `h1`s, but Docusaurus only accepts headings with a `level` of 2 or 3 (`h2`s and `h3`s only).

When refreshing a page, I'm taking a guess that Docusaurus's initial value of currentToc (which is pulled from docusaurus's `toc` variable) is the TOC of whatever the newest version of dbt is. 
This might be because we are not using their versioning but our own custom implementation of versioning. So in the case of the `The "file" or "fqn" method` section on `/reference/node-selection/methods`, it was checking the `currentToc` with a length of `15` to `updated`, which was missing the `The "file" or "fqn" method` data BUT included the `h1` data for `Methods` which also made the length `15`.

So, by strict checking the content itself with `if(currentToc !== updated)`, this check will trigger setting the `currentToc` state to whatever `current` is, instead of doing nothing.

Reply:
Nice discovery and refactor of that logic, makes sense.